### ISSUE TITLE
Currently not possible to use `setAllowNewAccounts(false)` with email/link authentication

### ISSUE THREAD

**Issue by theHilikus** [firebase/FirebaseUI-Android#1762]
It's currently not possible to create a user with the firebase admin sdk (or any other way outside of firebase UI) and have that user use email/link authentication. It seems that if you want to use email/link, you are forced to enable client registration (`setAllowNewAccounts(true)`), which is not always desirable.

I traced the code and the issue is that an externally created user always is created with a `password` provider id. This causes issues in `com.firebase.ui.auth.util.data.ProviderUtils#fetchSortedProviders`, which ends up throwing a developer error. 

This is related to #1735 but there it was mentioned something slightly different: once you enable email/link, previous users with email/password won't be able to log in, meaning to use email/link, ALL users must use it, which looks like a design decision that requires a new feature to be improved. The situation here looks to me more like a bug: a completely new user that has never logged in cannot login at all, even if it is the first user in a blank auth db or all the users were created after enabling email/link auth. Also, the fact that the application accepts an invalid configuration (`setAllowNewAccounts(false)` with `.enableEmailLinkSignIn()`, which will never work together) also makes it look more like a corner-case bug

 * FirebaseUI version: 6.2.0
  
### Steps to reproduce

- Create a user outside of firebaseUI
- Configure FirebaseUI with the following options

```
List<AuthUI.IdpConfig> providers = getAvailableProviders();
Intent loginIntent = AuthUI.getInstance()
                .createSignInIntentBuilder()
                .setAvailableProviders(providers)
                .build();
startActivityForResult(loginIntent, LOG_IN_REQUEST_CODE);

private List<AuthUI.IdpConfig> getAvailableProviders() {
        ActionCodeSettings actionCodeSettings = ActionCodeSettings.newBuilder()
                .setAndroidPackageName(BuildConfig.APPLICATION_ID, true, null)
                .setIOSBundleId(iOsBundleId)
                .setHandleCodeInApp(true)
                .setUrl("https://foo")
                .build();

        return Collections.singletonList(
                new EmailBuilder()
                        .setAllowNewAccounts(false)
                        .enableEmailLinkSignIn()
                        .setActionCodeSettings(actionCodeSettings)
                        .build());

    }
```
- Try to log in. It will fail

Finally, all this worked fine if using pure firebase-auth: you can create a user externally and log it in using email/link

**Comment by samtstern** [comment]
@theHilikus thanks for filing.  Sounds like there are two actions items here:

First we need to stop allowing this combination:
```
.setAllowNewAccounts(false)
.enableEmailLinkSignIn()
```
We should error out immediately as that's not going to work.

Can you explain a little more about this:
> Finally, all this worked fine if using pure firebase-auth: you can create a user externally and log it in using email/link

If you can help me recreate that scenario that would be really useful.

**Comment by theHilikus** [comment]
Hi @samtstern 
What i meant by the last bit is that if you create a user externally (e.g. using the [admin sdk](https://firebase.google.com/docs/auth/admin/manage-users#create_a_user)), you can still log it in using pure firebase-auth as [described here](https://firebase.google.com/docs/auth/android/email-link-auth#send_an_authentication_link_to_the_users_email_address) using the sample code that's there (`auth.sendSignInLinkToEmail`). But if you create another user and follow the procedure to log it in using firebaseUI instead of pure firebase-auth, it doesn't work. you hit the problem i mentioned. 

I hope that makes it clearer but let me know if you still have quesitons

**Comment by samtstern** [comment]
@theHilikus I know it's been a while but I finally sat down to address this and at first glance the UI is working for me.  This is what I see in the email link flow when allow new accounts is false:

![Screen Shot 2020-07-20 at 2 04 23 PM](https://user-images.githubusercontent.com/8466666/87940760-18d14980-ca68-11ea-8a29-bbf22dd1e8ab.png)

Do you see something else?

**Comment by theHilikus** [comment]
Hi @samtstern 
I see that as well. the problem is not when the user doesn't exist. it is when the user DOES exist but it was not created by firebaseUI, so `.setAllowNewAccounts` is set to false so that users can never be created by firebaseUI but it does need to exist to reproduce the bug

I create the user via the firebase admin sdk

Conceptually, think of a by-inventation-only system. An op has to create your account before you can log in; you cannot signup by yourself. This is the workflow that is currently not possible: the combination of email-link signin with setAllowNewAccounts = false

**Comment by theHilikus** [comment]
btw, I fixed this issue in my personal fork. I never opened a PR because i don't understand the whole project enough to guarantee that there are no unintended regressions, but if we can fix this over here that would be great

https://github.com/theHilikus/FirebaseUI-Android/commit/ce69ff9a8f84a85b33777df587fc3e1da7a2c3ed

The logic in my fix is, if you explicitly configure your builder to use email-link (i.e. enableEmailLinkSignIn()) but the provider only has email-password, force-add the email-link provider as the first option

**Comment by samtstern** [comment]
@theHilikus thanks for sharing that commit!  I tested it out and it seems fine, the worst case is a `DEVELOPER_ERROR` if you try and do something wrong which is fine with me.

I'm gonna patch that in.

**Comment by samtstern** [comment]
This issue is fixed and released in version `6.3.0`.
